lib/python/ensembl/brc4/runnable/dump_stable_ids.py

```diff
@@ -0,0 +1,468 @@
+        line_list = []
+
+        for pair in self.pairs:
+            line = [
+                pair["old_id"],
+                pair["new_id"],
+                name,
+                release,
+                date,
+            ]
+            line_list.append("\t".join(line))
+        return line_list
+    
+    @staticmethod
+    def clean_set(this_list: Set) -> Set:
+        """Removes any empty elements from a list.
+        
+        Args:
+            this_list: list of items, so of which can be empty/None.
+        
+        Returns:
+            The cleaned list.
+            
+        """
+        return set([id for id in this_list if id])
+    
+    def add_from(self, from_id: str) -> None:
+        """Store an id in the from_set."""
+        if from_id:
+            self.from_set.add(from_id)
+    
+    def add_to(self, to_id: str) -> None:
+        """Store an id in the from_set."""
+        if to_id:
+            self.to_set.add(to_id)
+    
+    def set_release(self, release: str) -> None:
+        self.release = release
+    
+    def set_date(self, date: datetime) -> None:
+        self.date = date
+    
+    def add_pair(self, pair: Dict) -> None:
+        """Keeps a record of this pair.
+        
+        Args:
+            pair: Dictionary of pairs to record, with keys "old_id" and "new_id".
+            
+        Raises:
+            ValueError: When no-empty value is provided for either "old_id" or "new_id".
+        
+        """
+        if "old_id" in pair and pair["old_id"] == None:
+            pair["old_id"] = ""
+        if "new_id" in pair and pair["new_id"] == None:
+            pair["new_id"] = ""
+        if not pair["old_id"] and not pair["new_id"]:
+            raise ValueError(f"Expected at least one value in the given pair {pair}")
+        self.pairs.append(pair)
+    
+    def get_full_release(self) -> str:
+        """Returns the expanded release name, pre-BRC4 or `BRC4 = build`."""
+        release = self.release
+        date = self.date
+
+        if date and date > BRC4_START_DATE:
+            release = f"build {release}"
+        else:
+            release = f"pre-BRC4 {release}"
+        
+        return release
+
+    def _name_event(self) -> None:
+        """Identify the event name based on the old vs new id lists."""
+        if not self.from_set and len(self.to_set) == 1:
+            self.name = "new"
+        elif not self.to_set and len(self.from_set) == 1:
+            self.name = "deletion"
+        elif len(self.from_set) == 1 and len(self.to_set) == 1:
+            self.name = "change"
+        elif len(self.from_set) == 1 and len(self.to_set) > 1:
+            self.name = "split"
+        elif len(self.from_set) > 1 and len(self.to_set) == 1:
+            self.name = "merge"
+        elif len(self.from_set) > 1 and len(self.to_set) > 1:
+            self.name = "mixed"
+        else:
+            raise UnsupportedEvent(f"Event {self.from_set} to {self.to_set} is not supported")
```

Comment:
I like this bit quite a lot.

Reply:
In principle it should never be invoked, but that's a safety in case some weird event gets added in the future...